text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+2. They are too powerful. Extensions are implemented via arbitrary Rust code, which is essential for encoding features that cannot be represented via Cedar expressions (such as IP Addresses), but opens the door for a wide range of bugs/design issues. It’s trivial to design an extension that is difficult to validate and/or logically encode for analysis. Problematically, extension functions can potentially exhibit non-determinism, non-termination, or non-linear performance; interact with the operating system; or violate memory safety. This raises the code review burden when considering an extension function's implementation.
+
+In contrast, functions written as simple abstractions over Cedar expressions, which themselves cannot call other Cedar functions, have none of these problems. They naturally inherit the properties of Cedar expressions. Cedar functions are guaranteed to terminate and be deterministic. Since they are compositions of Cedar expressions, it’s easy to validate and analyze them.
+
+## Detailed Design
+
+### Function declarations
+
+This RFC adds a new top level form to Cedar policysets: the function declaration.
+
+A Cedar function declaration is composed of three elements:
+
+1. A name, which is a valid (possibly namespaced) identifier.
+2. A list of parameters, each of which is a non-namespaced identifier.
+3. A body, which is a Cedar expression.
+    1. The body may contain (non-namespaced) variables, drawn from the function's parameters.
+    2. The body may not contain calls to other functions.
+
+
+Structurally, a declaration is written like this:
+
+```
+function name(param1, param2) {
+  body
+};
+```
+Standard Cedar variables (`principal`, `action`, `resource`, and `context`) are *not* considered bound with the body. 
+(Following the principal of macro hygiene)
+Use of an unbound variable in the body is a syntax error.
+A parameter list may not declare the same variable twice, and may not list any standard Cedar variables.
+An unused variable is a syntax warning.
+Function and variable names share the same namesapce, with standard lexical scoping.
+Inside of a function, any function application (see below) that does not resolve to an extension function or built-in operation is an error. In other words, Cedar functions are not permitted to call other Cedar functions.
+
+### Function applications (a.k.a. function calls)
+
+A function application has the same syntax as an extension function constructor application. In particular, a function application is composed of two elements:
+
+1. The function name (potentially namespaced)
+2. A comma separated list of arguments
+
+Here is an example:
+
+```
+foo(1,2, principal.name)
+```
+
+Function arguments are eagerly evaluated (i.e., call-by-value), as with extension functions today.
```

Comment:
We say the say below we will just inline the function body at each place it's called, when doing validation. If we are doing that, then we are essentially doing lazy evaluation, not eager evaluation. This could lead to validation soundness issues. In particular:
```
function drop(a,b) { a };
permit(principal,action,resource) 
when { drop(true,1+"hello") };
```
This policy will validate because once we've inlined the function we'd have
```
permit(principal,action,resource) 
when { true };
```
But if we actually do CBV evaluation then this policy will fail because `1+"hello"` will fail. We can probably solve this problem by validating the argument expressions of a function call individually, in addition to validating the entire policy after inlining.

But then there's also the problem with analysis: Our logical encoding would have to do CBV to be consistent with the actual evaluator, rather than just inlining, or else we need to prove that eager-eval(e) = lazy-eval(e) for all validated e.

For my money, I'd vote to do lazy evaluation and calling these things macros, replacing the `function` keyword with `def`. This allows users to define more expressive language features. For example:
```
def implies(e1,e1) {
  !e1 || (e1 && e2)
}
permit(principal,action,resource) when {
  implies(principal has attr,
          resource has attr && principal.attr == resource.attr)
};
```
The above won't work with CBV because the second expression to the call to `implies` will fail if `principal.attr` does not exist.

Reply:
Thought about this some and I agree completely, going to update the RFC

Reply:
Per Mike's comment, with macros, we could also relax the restriction that the body cannot reference Cedar variables.